BaragonService/src/main/java/com/hubspot/baragon/service/elb/ApplicationLoadBalancer.java

```diff
@@ -29,23 +62,413 @@ public ApplicationLoadBalancer(Optional<ElbConfiguration> configuration,
+        RegisterTargetsRequest registerTargetsRequest = new RegisterTargetsRequest()
+            .withTargetGroupArn(targetGroup.getTargetGroupArn())
+            .withTargets(targetDescriptions);
+        elbClient.registerTargets(registerTargetsRequest);
+        LOG.info("Registered targets {} onto target group {}", targetDescriptions, targetGroup);
+      } catch (AmazonClientException acexn) {
+        LOG.error("Failed to register targets {} onto target group {}", targetDescriptions, targetGroup);
+        exceptionNotifier.notify(acexn, ImmutableMap.of(
+            "targets", targetDescriptions.toString(),
+            "targetGroup", targetGroup.toString()));
+      }
+    }
+  }
+
+  private boolean agentShouldRegisterInTargetGroup(String baragonAgentInstanceId, TargetGroup targetGroup, Collection<TargetDescription> targets) {
+    boolean shouldRegister = true;
+    for (TargetDescription target : targets) {
+      if (target.getId().equals(baragonAgentInstanceId)) {
+        shouldRegister = false;
+      }
+    }
+    return shouldRegister;
+  }
+
+  private Collection<String> getSubnetsFromLoadBalancer(LoadBalancer loadBalancer) {
+    List<AvailabilityZone> availabilityZones = loadBalancer.getAvailabilityZones();
+    Set<String> subnetIds = new HashSet<>();
+
+    for (AvailabilityZone availabilityZone : availabilityZones) {
+      subnetIds.add(availabilityZone.getSubnetId());
+    }
+
+    return subnetIds;
+  }
+
+  private Optional<TargetGroup> guaranteeTargetGroupFor(BaragonGroup baragonGroup) {
+    DescribeTargetGroupsRequest targetGroupsRequest = new DescribeTargetGroupsRequest()
+        .withNames(baragonGroup.getName());
+    List<TargetGroup> targetGroups = elbClient.describeTargetGroups(targetGroupsRequest)
+        .getTargetGroups();
+    if (targetGroups.isEmpty()) {
+      LOG.info("No target group set up for BaragonGroup {}. Skipping.", baragonGroup);
+      return Optional.absent();
+    } else {
+      return Optional.of(targetGroups.get(0));
+    }
+  }
+
+  private Collection<TargetDescription> targetsInTargetGroup(TargetGroup targetGroup) {
+    DescribeTargetHealthRequest targetHealthRequest = new DescribeTargetHealthRequest()
+        .withTargetGroupArn(targetGroup.getTargetGroupArn());
+    List<TargetHealthDescription> targetGroupsResult = elbClient
+        .describeTargetHealth(targetHealthRequest)
+        .getTargetHealthDescriptions();
+
+    Collection<TargetDescription> targetDescriptions = new HashSet<>();
+    for (TargetHealthDescription targetHealthDescription : targetGroupsResult) {
+      targetDescriptions.add(targetHealthDescription.getTarget());
+    }
+
+    return targetDescriptions;
+  }
+
+  private Collection<TargetDescription> listOldTargets(BaragonGroup baragonGroup,
+                                                       Collection<TargetDescription> targetsOnGroup,
+                                                       Collection<BaragonAgentMetadata> knownAgents) {
```

Comment:
Be careful what you call 'known' for agents. There is a concept of `BaragonKnownAgentMetadata` which is the agent metadata along with a 'list seen' timestamp (stored in a separate zk node). However, what you are passing here is actually the regular agent metadata for the active agent (as stored in the leader-latch zk node)